autoscheduler/frontend/src/redux/reducers.ts

```diff
@@ -61,10 +65,120 @@ function courseCards(
   }
 }
 
+// sets whether the user is selecting busy time or not
+// in the future, this could also support preferred times
+function availabilityMode(
+  state: AvailabilityType = AvailabilityType.NONE, action: AvailabilityModeAction,
+): AvailabilityType {
+  if (action.type === SET_AVAILABILITY_MODE) { return action.mode; }
+  return state;
+}
+
+// stores all information about the user's availability
+function availability(
+  state: Availability[] = [], action: AvailabilityAction,
+): Availability[] {
+  switch (action.type) {
+    case ADD_AVAILABILITY:
+    case MERGE_AVAILABILITY: {
+      // helper functions to get the start and end times of an availability object in minutes
+      const getStart = (av: Availability): number => av.startTimeHours * 60 + av.startTimeMinutes;
+      const getEnd = (av: Availability): number => av.endTimeHours * 60 + av.endTimeMinutes;
+
+      // check for overlaps between the availability to be added and pre-existing ones
+      let overlapsFound = 0; // counts merging overlaps only
+      let overlapIdx = -1;
+
+      const initNewAv = action.type === ADD_AVAILABILITY
+        ? argsToAvailability(action.availability)
+        : state[state.length - 1];
+      let newAv = initNewAv;
+      return state.reduce<Availability[]>((avsList, oldAv, idx): Availability[] => {
+        // only counts as an overlap if they're on the same day of the week
+        if (oldAv.dayOfWeek === newAv.dayOfWeek) {
+          const avWithEarlierStart = getStart(oldAv) < getStart(newAv) ? oldAv : newAv;
+          const avWithLaterStart = avWithEarlierStart === oldAv ? newAv : oldAv;
+          // overlap detected
+          if (getEnd(avWithEarlierStart) >= getStart(avWithLaterStart)) {
+            // if they're the same type, merge them
+            if (oldAv.available === newAv.available) {
+              overlapsFound += 1;
+              const oldOverlapIdx = overlapIdx;
+              overlapIdx = idx;
+              const newEnd = Math.max(getEnd(oldAv), getEnd(newAv));
+              const newEndHrs = Math.floor(newEnd / 60);
+              const newEndMins = newEnd % 60;
+              newAv = {
+                available: oldAv.available,
+                dayOfWeek: oldAv.dayOfWeek,
+                startTimeHours: avWithEarlierStart.startTimeHours,
+                startTimeMinutes: avWithEarlierStart.startTimeMinutes,
+                endTimeHours: newEndHrs,
+                endTimeMinutes: newEndMins,
+              };
+              if (overlapsFound === 2) {
+                return avsList.slice(0, oldOverlapIdx)
+                  .concat(avsList.slice(oldOverlapIdx + 1))
+                  .concat(newAv);
+              }
+              return avsList.filter((av) => av !== oldAv).concat(newAv);
+            }
+            // if they're different types, then set the new one to the old one's borders
+            // TODO
```

Comment:
Hmm, that´s space for a possible future functionality that would allow the user to set times that they would *prefer* not to have classes. I'd like to leave the comment there to make it easier to figure out in case we ever get around to adding that.

Reply:
Yeah I figured as much. In that case the comment on 126 is fine, but I'd say remove the `TODO`, as I generally like to use it as a in-progress indicator (plus just don't like committing them in general, `pylint` throws errors for them on the backend too)